capsules/src/virtual_sha.rs

```diff
@@ -0,0 +1,178 @@
+    }
+
+    /// Add data to the sha IP.
+    /// All data passed in is fed to the SHA hardware block.
+    /// Returns the number of bytes written on success
+    fn add_data(
+        &self,
+        data: LeasableBuffer<'static, u8>,
+    ) -> Result<usize, (ErrorCode, &'static mut [u8])> {
+        // Check if any mux is enabled. If it isn't we enable it for us.
+        if self.mux.running.get() == false {
+            self.mux.running.set(true);
+            self.mux.running_id.set(self.id);
+            self.mux.sha.add_data(data)
+        } else if self.mux.running_id.get() == self.id {
+            self.mux.sha.add_data(data)
+        } else {
+            Err((ErrorCode::BUSY, data.take()))
+        }
+    }
+
+    /// Request the hardware block to generate a SHA
+    /// This doesn't return anything, instead the client needs to have
+    /// set a `hash_done` handler.
+    fn run(
+        &'a self,
+        digest: &'static mut [u8; L],
+    ) -> Result<(), (ErrorCode, &'static mut [u8; L])> {
+        // Check if any mux is enabled. If it isn't we enable it for us.
+        if self.mux.running.get() == false {
+            self.mux.running.set(true);
+            self.mux.running_id.set(self.id);
+            self.mux.sha.run(digest)
+        } else if self.mux.running_id.get() == self.id {
+            self.mux.sha.run(digest)
+        } else {
+            Err((ErrorCode::BUSY, digest))
+        }
+    }
+
+    /// Disable the SHA hardware and clear the keys and any other sensitive
+    /// data
+    fn clear_data(&self) {
+        if self.mux.running_id.get() == self.id {
+            self.mux.running.set(false);
+            self.mux.sha.clear_data()
+        }
+    }
+}
+
+impl<'a, A: digest::Digest<'a, L>, const L: usize> digest::Client<'a, L>
+    for VirtualMuxSha<'a, A, L>
+{
+    fn add_data_done(&'a self, result: Result<(), ErrorCode>, data: &'static mut [u8]) {
+        self.client
+            .map(move |client| client.add_data_done(result, data));
+    }
+
+    fn hash_done(&'a self, result: Result<(), ErrorCode>, digest: &'static mut [u8; L]) {
+        self.client
+            .map(move |client| client.hash_done(result, digest));
+    }
+}
+
+impl<'a, A: digest::Digest<'a, L> + digest::Sha256, const L: usize> digest::Sha256
+    for VirtualMuxSha<'a, A, L>
+{
+    fn set_mode_sha256(&self) -> Result<(), ErrorCode> {
+        // Check if any mux is enabled. If it isn't we enable it for us.
+        if self.mux.running.get() == false {
+            self.mux.running.set(true);
+            self.mux.running_id.set(self.id);
+            self.mux.sha.set_mode_sha256()
+        } else if self.mux.running_id.get() == self.id {
+            self.mux.sha.set_mode_sha256()
+        } else {
+            Err(ErrorCode::BUSY)
+        }
+    }
+}
+
+impl<'a, A: digest::Digest<'a, L> + digest::Sha384, const L: usize> digest::Sha384
+    for VirtualMuxSha<'a, A, L>
+{
+    fn set_mode_sha384(&self) -> Result<(), ErrorCode> {
```

Comment:
I don't think this is right: it's not virtualizing, it's exposing the underlying mux and failing if the current client can't gain control. This is the exact opposite of virtualization? The configuration needs to be stored in the VirtualMuxSha then applied when this particular virtual instance is allowed to perform an operation.

But more generally, how are requests queued? This is a key element of virtualization.